tests/std/tests/P0896R4_ranges_range_machinery/test.cpp

```diff
@@ -1886,6 +1886,48 @@ namespace unwrapped_begin_end {
     }
 } // namespace unwrapped_begin_end
 
+namespace closure {
+    // Verify that range adaptor closure capture with the proper value category
+
+    enum class path { none, lvalue, const_lvalue, rvalue, const_rvalue };
+
+    template <path allowed>
+    struct arg {
+        constexpr arg(arg&) {
+            static_assert(allowed == path::lvalue);

```

Comment:
`allowed` is a template parameter belonging to the `struct` and not this constructor, so I believe that compilers are permitted to eagerly diagnose here (and I'm slightly surprised that they aren't). Should this have a layer of indirection (`template <Path Allowed2 = Allowed> CONSTRUCTOR`)?

Reply:
Declarations of member functions are instantiated when the class is instantiated, but the bodies of member functions of a class template are instantiated only when they are called. If a compiler were to diagnose something in the body of this class member function _before_ its instantiated, it would have to do so in accordance with [temp.res.general]/6 which begins with "The validity of a template may be checked prior to any instantiation." Prior to _any_ instantiation of the class template, `allowed` is value-dependent and therefore fine. There is no facility for checking partially-instantiated members of class templates, indeed doing so would be counter to the design of class templates for which its intended that not every member function must be valid for every choice of class template argument.

(Simplified example: https://godbolt.org/z/W83zjYTnP.)